astropy/coordinates/attributes.py

```diff
@@ -99,7 +117,7 @@ def convert_input(self, value):
 
     def __get__(self, instance, frame_cls=None):
         if instance is None:
-            out = self.default
+            return self
```

Comment:
I think we are risking spending a lot of time discussing something that I feel is a nice improvement for users on the basis that it might not be ideal for data classes (which is not directly relevant since really `BaseFrame` has no need to become one). How about we make things slightly future-proof by using the fix we initially had for `MetaData`, i.e., here
```
return self.default if is_dataclass(frame_cls) else self
```
(with a `from dataclasses import is_dataclass` on top)

I would also suggest not to inherit from `property` for now -- let's try to get that fixed in `ipython`.

Reply:
Also need to do in ``__set__``

```python
if value is self:
    value = self.default
````

It's not brilliant for dataclasses since the default value on the class will be the descriptor, but when instantiated the descriptor instance will be converted to the correct value.

Reply:
> How about we make things slightly future-proof by using the fix we initially had for `MetaData`, i.e., here
> 
> ```
> return self.default if is_dataclass(frame_cls) else self
> ```

Done

> I would also suggest not to inherit from `property` for now -- let's try to get that fixed in `ipython`.

Reverted

> Also need to do in `__set__`
> 
> ```python
> if value is self:
>     value = self.default
> ```

Take a look at what I did, since I continue to block setting when not in a dataclass.  I used `object.__setattr__()` as was done in #15404 so that frozen dataclasses would work.